pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,296 @@
+)
+
+// ValidateFlowSchemaName validates name for flow-schema.
+var ValidateFlowSchemaName = apimachineryvalidation.NameIsDNSSubdomain
+
+// ValidatePriorityLevelConfigurationName validates name for priority-level-configuration.
+var ValidatePriorityLevelConfigurationName = apimachineryvalidation.NameIsDNSSubdomain
+
+var supportedDistinguisherMethods = sets.NewString(
+	string(flowcontrol.FlowDistinguisherMethodByNamespaceType),
+	string(flowcontrol.FlowDistinguisherMethodByUserType),
+)
+
+var supportedVerbs = sets.NewString(
+	flowcontrol.VerbAll,
+	"get",
+	"list",
+	"watch",
+	"create",
+	"update",
+	"delete",
+	"deletecollection",
+	"patch",
+)
+
+var supportedSubjectKinds = sets.NewString(
+	string(flowcontrol.SubjectKindServiceAccount),
+	string(flowcontrol.SubjectKindGroup),
+	string(flowcontrol.SubjectKindUser),
+)
+
+var supportedQueuingType = sets.NewString(
+	string(flowcontrol.PriorityLevelQueuingTypeQueueing),
+	string(flowcontrol.PriorityLevelQueuingTypeExempt),
+)
+
+// ValidateFlowSchema validates the content of flow-schema
+func ValidateFlowSchema(fs *flowcontrol.FlowSchema) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&fs.ObjectMeta, false, ValidateFlowSchemaName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateFlowSchemaSpec(&fs.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidateFlowSchemaStatus(&fs.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+// ValidateFlowSchemaSpec validates the content of flow-schema's spec
+func ValidateFlowSchemaSpec(spec *flowcontrol.FlowSchemaSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if spec.MatchingPrecedence < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("matchingPrecedence"), spec.MatchingPrecedence, "must be positive value"))
+	}
+	if spec.DistinguisherMethod != nil {
+		if !supportedDistinguisherMethods.Has(string(spec.DistinguisherMethod.Type)) {
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("distinguisherMethod").Child("type"), spec.DistinguisherMethod, supportedDistinguisherMethods.List()))
+		}
+	}
+	if len(spec.PriorityLevelConfiguration.Name) > 0 {
+		for _, msg := range ValidatePriorityLevelConfigurationName(spec.PriorityLevelConfiguration.Name, false) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("priorityLevelConfiguration").Child("name"), spec.PriorityLevelConfiguration.Name, msg))
+		}
+	} else {
+		allErrs = append(allErrs, field.Required(fldPath.Child("priorityLevelConfiguration").Child("name"), "must reference to a priority level"))
+	}
+	for i, rule := range spec.Rules {
+		allErrs = append(allErrs, ValidateFlowSchemaPolicyRuleWithSubjects(&rule, fldPath.Child("rules").Index(i))...)
+	}
+	return allErrs
+}
+
+// ValidateFlowSchemaPolicyRuleWithSubjects validates policy-rule-with-subjects object.
+func ValidateFlowSchemaPolicyRuleWithSubjects(rule *flowcontrol.PolicyRulesWithSubjects, fldPath *field.Path) field.ErrorList {
```

Comment:
(1) Ideally this function's name should be updated to match the change in the name of the type of its first parameter.

(2) Do we want subsidiary functions like this to be public?

Reply:
renamed function